graphql2/graphqlapp/expr.go

```diff
@@ -0,0 +1,306 @@
+package graphqlapp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"slices"
+	"strings"
+
+	"github.com/99designs/gqlgen/graphql"
+	"github.com/expr-lang/expr/ast"
+	"github.com/expr-lang/expr/parser"
+	"github.com/pkg/errors"
+	"github.com/target/goalert/graphql2"
+	"github.com/target/goalert/validation"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+)
+
+type Expr App
+
+func (a *App) Expr() graphql2.ExprResolver                    { return (*Expr)(a) }
+func (q *Query) Expr(context.Context) (*graphql2.Expr, error) { return &graphql2.Expr{}, nil }
+
+var errTooComplex = errors.New("expression is too complex")
+
+func gqlErrTooComplex(ctx context.Context) error {
+	return &gqlerror.Error{
+		Message: errTooComplex.Error(),
+		Path:    graphql.GetPath(ctx),
+		Extensions: map[string]interface{}{
+			"code": graphql2.ErrorCodeExprTooComplex,
+		},
+	}
+}
+
+var supportedOperators = []string{"==", "!=", "<", ">", "<=", ">=", "in"}
+
+func isID(n ast.Node) bool {
+	switch t := n.(type) {
+	case *ast.IdentifierNode:
+		return true
+	case *ast.MemberNode:
+		return !t.Method && isID(t.Node) && (isID(t.Property) || isLiteral(t.Property))
+	case *ast.ChainNode:
+		return isID(t.Node)
+	}
+
+	return false
+}
+
+func isLiteral(n ast.Node) bool {
+	switch t := n.(type) {
+	case *ast.StringNode:
+	case *ast.IntegerNode:
+	case *ast.BoolNode:
+	case *ast.FloatNode:
+	case *ast.UnaryNode:
+		return isLiteral(t.Node)
+	case *ast.ArrayNode:
+		for _, v := range t.Nodes {
+			if !isLiteral(v) {
+				return false
+			}
+		}
+	default:
+		return false
+	}
+
+	return true
+}
+
+func containsClause(n *ast.BinaryNode) (clause *graphql2.Clause, ok bool) {
+	call, ok := n.Left.(*ast.BuiltinNode)
+	if !ok {
+		return nil, false
+	}
+	un, ok := n.Right.(*ast.UnaryNode)
+	if !ok || un.Operator != "-" {
+		return nil, false
+	}
+	val, ok := un.Node.(*ast.IntegerNode)
+	if !ok || val.Value != 1 {
+		return nil, false
+	}
+
+	if call.Name != "indexOf" {
+		return nil, false
+	}
+	if len(call.Arguments) != 2 {
+		return nil, false
+	}
+
+	if !isID(call.Arguments[0]) {
+		return nil, false
+	}
+
+	if !isLiteral(call.Arguments[1]) {
+		return nil, false
+	}
+
+	var op string
+	switch n.Operator {
+	case "==":
```

Comment:
would it be helpful to define these string literals as constants as they show up several times?

Reply:
I think the extra indirection would likely hurt readability more than it would help, _but_ I think adding comments to this to explaining the _why_ would help, so I'll work on adding those